Provider v0.84.0: Crash in `apstra_datacenter_virtual_network` resource with Apstra 5.1

Reported OoB by a user:

```
apstra_datacenter_virtual_network.STORE-SB-314_vlan_id_314: Creating...
╷
│ Error: Plugin did not respond
│
│ The plugin encountered an error, and failed to respond to the plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-apstra_v0.84.0 plugin:
panic: runtime error: index out of range [0] with length 0
goroutine 42 [running]:
github.com/Juniper/terraform-provider-apstra/apstra/blueprint.(*DatacenterVirtualNetwork).Request(0xc000514fc0, {0x124c588, 0xc00045ec90}, 0xc000463898)
        github.com/Juniper/terraform-provider-apstra/apstra/blueprint/datacenter_virtual_network.go:604 +0x1125
github.com/Juniper/terraform-provider-apstra/apstra.(*resourceDatacenterVirtualNetwork).Create(0xc0002c6258, {0x124c588, 0xc00045ec90}, {{{{0x1255218, 0xc0005b7020}, {0xefdae0, 0xc0005b67b0}}, {0x1259040, 0xc00011d0e0}}, {{{0x1255218, ...}, ...}, ...}, ...}, ...)
        github.com/Juniper/terraform-provider-apstra/apstra/resource_datacenter_virtual_network.go:186 +0x3f1
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).CreateResource(0xc0003d6488, {0x124c588, 0xc00045ec90}, 0xc0005e9538, 0xc0005e9510)
        github.com/hashicorp/terraform-plugin-framework@v1.13.0/internal/fwserver/server_createresource.go:101 +0x578
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).ApplyResourceChange(0xc0003d6488, {0x124c588, 0xc00045ec90}, 0xc000462410, 0xc0005e9628)
        github.com/hashicorp/terraform-plugin-framework@v1.13.0/internal/fwserver/server_applyresourcechange.go:57 +0x4aa
github.com/hashicorp/terraform-plugin-framework/internal/proto6server.(*Server).ApplyResourceChange(0xc0003d6488, {0x124c588?, 0xc00045eba0?}, 0xc000462370)
        github.com/hashicorp/terraform-plugin-framework@v1.13.0/internal/proto6server/server_applyresourcechange.go:55 +0x38e
github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server).ApplyResourceChange(0xc0003dc0a0, {0x124c588?, 0xc00045e1e0?}, 0xc00054c1c0)
        github.com/hashicorp/terraform-plugin-go@v0.25.0/tfprotov6/tf6server/server.go:865 +0x3d0
github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_ApplyResourceChange_Handler({0x1053100, 0xc0003dc0a0}, {0x124c588, 0xc00045e1e0}, 0xc0005a4000, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.25.0/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:611 +0x1a6
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000416000, {0x124c588, 0xc00045e150}, {0x12569e0, 0xc0000f1860}, 0xc000464000, 0xc0003c3ad0, 0x19491f8, 0x0)
        google.golang.org/grpc@v1.67.1/server.go:1394 +0xe49
google.golang.org/grpc.(*Server).handleStream(0xc000416000, {0x12569e0, 0xc0000f1860}, 0xc000464000)
        google.golang.org/grpc@v1.67.1/server.go:1805 +0xe8b
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.67.1/server.go:1029 +0x8b
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 8
        google.golang.org/grpc@v1.67.1/server.go:1040 +0x125

Error: The terraform-provider-apstra_v0.84.0 plugin crashed!
This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

--- a/apstra/blueprint/datacenter_virtual_network.go
+++ b/apstra/blueprint/datacenter_virtual_network.go
@@ -964,7 +964,7 @@ func (o DatacenterVirtualNetwork) ValidateConfigBindingsReservation(ctx context.
 		}
 		if binding.VlanId.IsNull() {
 			invalidConfigDueToNullVlan = true
-			continue // todo: should this be 'break' instead?
+			break
 		}
 		reservedVlanIds[binding.VlanId.ValueInt64()] = struct{}{}
 	}

--- a/apstra/blueprint/datacenter_virtual_network.go
+++ b/apstra/blueprint/datacenter_virtual_network.go
@@ -12,6 +12,7 @@ import (
 	apiversions "github.com/Juniper/terraform-provider-apstra/apstra/api_versions"
 	"github.com/Juniper/terraform-provider-apstra/apstra/compatibility"
 	"github.com/Juniper/terraform-provider-apstra/apstra/constants"
+	"github.com/Juniper/terraform-provider-apstra/apstra/design"
 	apstraregexp "github.com/Juniper/terraform-provider-apstra/apstra/regexp"
 	"github.com/Juniper/terraform-provider-apstra/apstra/utils"
 	apstravalidator "github.com/Juniper/terraform-provider-apstra/apstra/validator"
@@ -42,6 +43,7 @@ type DatacenterVirtualNetwork struct {
 	Vni                     types.Int64  `tfsdk:"vni"`
 	HadPriorVniConfig       types.Bool   `tfsdk:"had_prior_vni_config"`
 	ReserveVlan             types.Bool   `tfsdk:"reserve_vlan"`
+	ReservedVlanId          types.Int64  `tfsdk:"reserved_vlan_id"`
 	Bindings                types.Map    `tfsdk:"bindings"`
 	DhcpServiceEnabled      types.Bool   `tfsdk:"dhcp_service_enabled"`
 	IPv4ConnectivityEnabled types.Bool   `tfsdk:"ipv4_connectivity_enabled"`
@@ -106,6 +108,10 @@ func (o DatacenterVirtualNetwork) DataSourceAttributes() map[string]dataSourceSc
 				"use the same VLAN ID. This option reserves the VLAN fabric-wide, even on switches to "+
 				"which the Virtual Network has not yet been deployed.", enum.VnTypeVxlan),
 			Computed: true,
+		},
+		"reserved_vlan_id": dataSourceSchema.Int64Attribute{
+			MarkdownDescription: "Reserved VLAN ID, if any.",
+			Computed:            true,
 		},
 		"bindings": dataSourceSchema.MapNestedAttribute{
 			MarkdownDescription: "Details availability of the virtual network on leaf and access switches",
@@ -212,10 +218,12 @@ func (o DatacenterVirtualNetwork) DataSourceFilterAttributes() map[string]dataSo
 			Computed:            true,
 		},
 		"reserve_vlan": dataSourceSchema.BoolAttribute{
-			MarkdownDescription: fmt.Sprintf("For use only with `%s` type Virtual networks when all `bindings` "+
-				"use the same VLAN ID. This option reserves the VLAN fabric-wide, even on switches to "+
-				"which the Virtual Network has not yet been deployed.", enum.VnTypeVxlan),
-			Optional: true,
+			MarkdownDescription: "Selects only virtual networks with the *Reserve across blueprint* box checked.",
+			Optional:            true,
+		},
+		"reserved_vlan_id": dataSourceSchema.StringAttribute{
+			MarkdownDescription: "Selects only virtual networks with the *Reserve across blueprint* box checked and this value selected.",
+			Optional:            true,
 		},
 		"bindings": dataSourceSchema.MapNestedAttribute{
 			MarkdownDescription: "Not applicable in filter context. Ignore.",
@@ -384,6 +392,21 @@ func (o DatacenterVirtualNetwork) ResourceAttributes() map[string]resourceSchema
 						false,
 					),
 				),
+				apstravalidator.AlsoRequiresNOf(1,
+					path.MatchRoot("bindings"),
+					path.MatchRoot("reserved_vlan_id"),
+				),
+			},
+		},
+		"reserved_vlan_id": resourceSchema.Int64Attribute{
+			MarkdownDescription: "Used to specify the reserved VLAN ID without specifying any *bindings*.",
+			Optional:            true,
+			Computed:            true,
+			Validators: []validator.Int64{
+				apstravalidator.ForbiddenWhenValueIs(path.MatchRoot("reserve_vlan"), types.BoolNull()),
+				apstravalidator.ForbiddenWhenValueIs(path.MatchRoot("reserve_vlan"), types.BoolValue(false)),
+				int64validator.ConflictsWith(path.MatchRoot("bindings")),
+				int64validator.Between(design.VlanMin, design.VlanMax),
 			},
 		},
 		"bindings": resourceSchema.MapNestedAttribute{
@@ -601,7 +624,11 @@ func (o *DatacenterVirtualNetwork) Request(ctx context.Context, diags *diag.Diag
 
 	var reservedVlanId *apstra.Vlan
 	if o.ReserveVlan.ValueBool() {
-		reservedVlanId = vnBindings[0].VlanId
+		if !o.ReservedVlanId.IsNull() {
+			reservedVlanId = utils.ToPtr(apstra.Vlan(o.ReservedVlanId.ValueInt64()))
+		} else {
+			reservedVlanId = vnBindings[0].VlanId
+		}
 	}
 
 	var ipv4Subnet, ipv6Subnet *net.IPNet
@@ -687,6 +714,11 @@ func (o *DatacenterVirtualNetwork) LoadApiData(ctx context.Context, in *apstra.V
 	o.IPv4ConnectivityEnabled = types.BoolValue(in.Ipv4Enabled)
 	o.IPv6ConnectivityEnabled = types.BoolValue(in.Ipv6Enabled)
 	o.ReserveVlan = types.BoolValue(in.ReservedVlanId != nil)
+	if in.ReservedVlanId == nil {
+		o.ReservedVlanId = types.Int64Null()
+	} else {
+		o.ReservedVlanId = types.Int64Value(int64(*in.ReservedVlanId))
+	}
 	if in.Ipv4Subnet == nil {
 		o.IPv4Subnet = types.StringNull()
 	} else {
@@ -750,6 +782,13 @@ func (o *DatacenterVirtualNetwork) Query(resultName string) apstra.QEQuery {
 		nodeAttributes = append(nodeAttributes, apstra.QEEAttribute{
 			Key:   "reserved_vlan_id",
 			Value: apstra.QENone(!o.ReserveVlan.ValueBool()),
+		})
+	}
+
+	if !o.ReservedVlanId.IsNull() {
+		nodeAttributes = append(nodeAttributes, apstra.QEEAttribute{
+			Key:   "reserved_vlan_id",
+			Value: apstra.QEIntVal(o.ReservedVlanId.ValueInt64()),
 		})
 	}
 

--- a/apstra/resource_datacenter_virtual_network_integration_test.go
+++ b/apstra/resource_datacenter_virtual_network_integration_test.go
@@ -26,14 +26,16 @@ import (
 const (
 	resourceDatacenterVirtualNetworkTemplateHCL = `
 resource %q %q {
-  blueprint_id    = %q
-  name            = %q
-  description     = %s
-  type            = %s
-  vni             = %s
-  routing_zone_id = %s
-  l3_mtu          = %s
-  bindings        = %s
+  blueprint_id     = %q
+  name             = %q
+  description      = %s
+  type             = %s
+  vni              = %s
+  routing_zone_id  = %s
+  l3_mtu           = %s
+  bindings         = %s
+  reserve_vlan     = %s
+  reserved_vlan_id = %s
 }
 `
 	resourceDatacenterVirtualNetworkTemplateBindingHCL = `
@@ -45,14 +47,16 @@ resource %q %q {
 )
 
 type resourceDatacenterVirtualNetworkTemplate struct {
-	blueprintId   apstra.ObjectId
-	name          string
-	description   string
-	vnType        string
-	vni           *int
-	routingZoneId apstra.ObjectId
-	l3Mtu         *int
-	bindings      []resourceDatacenterVirtualNetworkTemplateBinding
+	blueprintId    apstra.ObjectId
+	name           string
+	description    string
+	vnType         string
+	vni            *int
+	routingZoneId  apstra.ObjectId
+	l3Mtu          *int
+	bindings       []resourceDatacenterVirtualNetworkTemplateBinding
+	reserveVlan    *bool
+	reservedVlanId *int
 }
 
 func (o resourceDatacenterVirtualNetworkTemplate) render(rType, rName string) string {
@@ -78,6 +82,8 @@ func (o resourceDatacenterVirtualNetworkTemplate) render(rType, rName string) st
 		stringOrNull(o.routingZoneId.String()),
 		intPtrOrNull(o.l3Mtu),
 		bindings.String(),
+		boolPtrOrNull(o.reserveVlan),
+		intPtrOrNull(o.reservedVlanId),
 	)
 }
 
@@ -593,6 +599,31 @@ func TestAccDatacenterVirtualNetwork(t *testing.T) {
 						description:   acctest.RandString(6),
 						vnType:        enum.VnTypeVxlan.String(),
 						routingZoneId: szId,
+					},
+				},
+			},
+		},
+		"no_bindings_reserved_vlan_id": {
+			apiVersionConstraints: compatibility.VnEmptyBindingsOk,
+			steps: []testStep{
+				{
+					config: resourceDatacenterVirtualNetworkTemplate{
+						blueprintId:    bp.Id(),
+						name:           acctest.RandString(6),
+						vnType:         enum.VnTypeVxlan.String(),
+						routingZoneId:  szId,
+						reserveVlan:    utils.ToPtr(true),
+						reservedVlanId: utils.ToPtr(1100),
+					},
+				},
+				{
+					config: resourceDatacenterVirtualNetworkTemplate{
+						blueprintId:    bp.Id(),
+						name:           acctest.RandString(6),
+						vnType:         enum.VnTypeVxlan.String(),
+						routingZoneId:  szId,
+						reserveVlan:    utils.ToPtr(true),
+						reservedVlanId: utils.ToPtr(1101),
 					},
 				},
 			},

--- a/apstra/validator/also_requires_n_of.go
+++ b/apstra/validator/also_requires_n_of.go
@@ -43,6 +43,10 @@ func (o AlsoRequiresNOfValidator) MarkdownDescription(_ context.Context) string 
 }
 
 func (o AlsoRequiresNOfValidator) Validate(ctx context.Context, req AlsoRequiresNOfValidatorRequest, resp *AlsoRequiresNOfValidatorResponse) {
+	if req.ConfigValue.IsNull() {
+		return
+	}
+
 	expressions := req.PathExpression.MergeExpressions(o.PathExpressions...)
 	for i := range expressions {
 		expressions[i] = expressions[i].Resolve()

--- a/docs/data-sources/datacenter_virtual_network.md
+++ b/docs/data-sources/datacenter_virtual_network.md
@@ -58,6 +58,7 @@ locals {
 - `ipv6_virtual_gateway_enabled` (Boolean) Controls and indicates whether the IPv6 gateway within the Virtual Network is enabled.
 - `l3_mtu` (Number) L3 MTU used by the L3 switch interfaces participating in the Virtual Network. Requires Apstra 4.2 or later.
 - `reserve_vlan` (Boolean) For use only with `vxlan` type Virtual networks when all `bindings` use the same VLAN ID. This option reserves the VLAN fabric-wide, even on switches to which the Virtual Network has not yet been deployed.
+- `reserved_vlan_id` (Number) Reserved VLAN ID, if any.
 - `routing_zone_id` (String) Routing Zone ID (only applies when `type == vxlan`
 - `type` (String) Virtual Network Type
 - `vni` (Number) EVPN Virtual Network ID to be associated with this Virtual Network.

--- a/docs/data-sources/datacenter_virtual_networks.md
+++ b/docs/data-sources/datacenter_virtual_networks.md
@@ -78,7 +78,8 @@ Optional:
 - `ipv6_virtual_gateway_enabled` (Boolean) Controls and indicates whether the IPv6 gateway within the Virtual Network is enabled.
 - `l3_mtu` (Number) L3 MTU used by the L3 switch interfaces participating in the Virtual Network. Requires Apstra 4.2 or later.
 - `name` (String) Virtual Network Name
-- `reserve_vlan` (Boolean) For use only with `vxlan` type Virtual networks when all `bindings` use the same VLAN ID. This option reserves the VLAN fabric-wide, even on switches to which the Virtual Network has not yet been deployed.
+- `reserve_vlan` (Boolean) Selects only virtual networks with the *Reserve across blueprint* box checked.
+- `reserved_vlan_id` (String) Selects only virtual networks with the *Reserve across blueprint* box checked and this value selected.
 - `routing_zone_id` (String) Routing Zone ID (required when `type == vxlan`)
 - `type` (String) Virtual Network Type
 - `vni` (Number) EVPN Virtual Network ID to be associated with this Virtual Network.
@@ -114,7 +115,8 @@ Optional:
 - `ipv6_virtual_gateway_enabled` (Boolean) Controls and indicates whether the IPv6 gateway within the Virtual Network is enabled.
 - `l3_mtu` (Number) L3 MTU used by the L3 switch interfaces participating in the Virtual Network. Requires Apstra 4.2 or later.
 - `name` (String) Virtual Network Name
-- `reserve_vlan` (Boolean) For use only with `vxlan` type Virtual networks when all `bindings` use the same VLAN ID. This option reserves the VLAN fabric-wide, even on switches to which the Virtual Network has not yet been deployed.
+- `reserve_vlan` (Boolean) Selects only virtual networks with the *Reserve across blueprint* box checked.
+- `reserved_vlan_id` (String) Selects only virtual networks with the *Reserve across blueprint* box checked and this value selected.
 - `routing_zone_id` (String) Routing Zone ID (required when `type == vxlan`)
 - `type` (String) Virtual Network Type
 - `vni` (Number) EVPN Virtual Network ID to be associated with this Virtual Network.

--- a/docs/resources/datacenter_virtual_network.md
+++ b/docs/resources/datacenter_virtual_network.md
@@ -70,6 +70,7 @@ resource "apstra_datacenter_virtual_network" "test" {
 - `ipv6_virtual_gateway_enabled` (Boolean) Controls and indicates whether the IPv6 gateway within the Virtual Network is enabled. Requires `ipv6_connectivity_enabled` to be `true`
 - `l3_mtu` (Number) L3 MTU used by the L3 switch interfaces participating in the Virtual Network. Must be an even number between 1280 and 9216. Requires Apstra 4.2.0 or later.
 - `reserve_vlan` (Boolean) For use only with `vxlan` type Virtual networks when all `bindings` use the same VLAN ID. This option reserves the VLAN fabric-wide, even on switches to which the Virtual Network has not yet been deployed. The only accepted values is `true`.
+- `reserved_vlan_id` (Number) Used to specify the reserved VLAN ID without specifying any *bindings*.
 - `routing_zone_id` (String) Routing Zone ID (required when `type == vxlan`
 - `type` (String) Virtual Network Type
 - `vni` (Number) EVPN Virtual Network ID to be associated with this Virtual Network.  When omitted, Apstra chooses a VNI from the Resource Pool [allocated](../resources/datacenter_resource_pool_allocation) to role `vni_virtual_network_ids`.

